streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java

```diff
@@ -349,6 +360,119 @@ public String queryableStoreName() {
+
+        final ProcessorSupplier<K, Change<V>> suppressionSupplier =
+            () -> new KTableSuppressProcessor<>(buildSuppress(suppressImpl));
+
+        final ProcessorParameters<K, Change<V>> processorParameters = new ProcessorParameters<>(
+            suppressionSupplier,
+            name
+        );
+
+        final ProcessorGraphNode<K, Change<V>> node = new ProcessorGraphNode<>(name, processorParameters, false);
+
+        builder.addGraphNode(this.streamsGraphNode, node);
+
+        return new KTableImpl<K, S, V>(
+            builder,
+            name,
+            suppressionSupplier,
+            keySerde,
+            valSerde,
+            Collections.singleton(this.name),
+            null,
+            false,
+            node
+        );
+    }
+
+    @SuppressWarnings("unchecked") // these casts are silly
+    private SuppressImpl<K, V> buildSuppress(final SuppressImpl<K, V> suppress) {
+        if (suppress.isFinalResultsSuppression()) {
+            // This throws an exception if the computation is not windowed
+            final long grace = findAndVerifyWindowGrace(streamsGraphNode, "");
+
+            // so we know we can cast the key to Windowed
+            final SuppressImpl<? extends Windowed, V> windowedSuppress = (SuppressImpl<? extends Windowed, V>) suppress;
+
+            final Suppress<? extends Windowed, V> finalResultsSuppression = SuppressImpl.buildFinalResultsSuppression(
+                Duration.ofMillis(grace),
+                windowedSuppress
+            );
+
+            //
+            return (SuppressImpl<K, V>) finalResultsSuppression;
+        } else {
+            return suppress;
+        }
+    }
+
+    private static long findAndVerifyWindowGrace(final StreamsGraphNode streamsGraphNode, final String chain) {
+        // error base case: we traverses off the end of the graph without finding a window definition
+        if (streamsGraphNode == null) {
+            throw new IllegalArgumentException(
+                "Window close time is only defined for windowed computations. Got [" + chain + "]."
+            );
+        }
+        // base case: return if this node defines a grace period.
+        {
+            final Long gracePeriod = extractGracePeriod(streamsGraphNode);
+            if (gracePeriod != null) {
+                return gracePeriod;
+            }
+        }
+
+        final String newChain = chain.equals("") ? streamsGraphNode.nodeName() : streamsGraphNode.nodeName() + "->" + chain;
+
+        // recursive case: all parents must define a grace period, and they must match.
+        long firstParentGrace = -1;
+        for (final StreamsGraphNode parentNode : streamsGraphNode.parentNodes()) {
+            final long parentGrace = findAndVerifyWindowGrace(parentNode, newChain);
+            if (firstParentGrace == -1) {
+                firstParentGrace = parentGrace;
+            } else {
+                if (parentGrace != firstParentGrace) {
```

Comment:
Actually today we have undefined operator for windowed-table / windowed-table join at all. But the logic itself looks good :P

Reply:
I see. Well, it's permitted by the API, so I'd rather keep the sanity check.

Actually, it seems like the only problem with further manipulations of windowed tables is that we won't know to use a windowed store (as by that point, we only know it's a KTable). Might be worth revisiting this at some point...